tokio/src/sync/mpsc/bounded.rs

```diff
@@ -230,6 +230,82 @@ impl<T> Receiver<T> {
         poll_fn(|cx| self.chan.recv(cx)).await
     }
 
+    /// Receives the next values for this receiver and extends `buffer`.
+    ///
+    /// This method extends `buffer` by no more than a fixed number
+    /// of values as specified by `limit`.  If `limit` is zero,
+    /// then a default is used. The return value is the number
+    /// of values added to `buffer`.
```

Comment:
If the limit is zero, then I think it would be okay to just return 0 immediately.

Reply:
If the caller wants a no-op, they can instead simply not-call the method, instead of calling with a zero limit.

Furthermore, immediately returning 0 means that this behavior becomes unique to a single value of `limit`; all other values of `limit` only return 0 if the channel is closed and no values are pending.  So it breaks the API's guarantee for a case of no practical value to someone seeking to receive messages on the channel.

I considered the following possible ways to handle a 0 value for limit:

- The fail-fast approach -- `assert!(limit > 0);`.
- Have it so that `limit=0` acts the same way as `limit=usize::max_value()`.
- Have it so that `limit=0` results in some reasonable default behavior, e.g., retrieve 2 or 8 messages or BLOCK_CAP.  I liked BLOCK_CAP because it corresponded to the internal chunking used.

Reply:
Hmm. I think immediately returning 0 is the least surprising behavior. After all, you asked to receive zero messages, and you got zero messages.

Reply:
Recall, you wrote:

>  Perhaps we can swap these paragraphs, and add something to the paragraph about 0 along the lines of "this method will never return 0 if the channel is not closed" or "if the channel is not closed, then this method will never return 0".

Nonetheless, I see that the same no-op convention is used in unistd.h's `read` function... and evidence of users running into the types of bugs I anticipate:
https://stackoverflow.com/a/3074217

If this is acceptable, I'll update the handling and the documentation.

Reply:
If what is acceptable?

I mainly see people be confused about length zero reads when they do this:
```rs
let mut vec = Vec::with_capacity(1024);
io.read(&mut vec);
```
and are surprised because `&mut vec` becomes a slice of length zero.

But that doesn't happen in our case.

Reply:
Understood regarding `io.read(&mut vec);` -- here the issue is resolved as we take a Vec.

The behavior of `recv_many(buf, 0)` is similar to calling [`rd.take(0)`](https://docs.rs/tokio/latest/tokio/io/trait.AsyncReadExt.html#method.take).  I am unclear when in an async context `take(0)` is ever useful -- but it is allowed -- so at least `recv_many(buf, 0)` is no worse.

Contrived example adapted from [Example: split a file into chunks](https://geekingfrog.com/blog/post/getting-things-done-with-async)
```
#[tokio::test]
async fn take_zero() {
    use tokio::fs::File;
    use std::path::PathBuf;
    for chunk_size in [0, 1, 1024] {
        let mut input_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        input_path.push("tests/io_take.rs");
        let input_file = File::open(input_path).await.unwrap();
        let mut output_path = PathBuf::from(env!("CARGO_TARGET_TMPDIR"));
        output_path.push("takezero");
        if output_path.exists() {
            tokio::fs::remove_file(&output_path).await.unwrap();
        }
        assert!(!output_path.exists());

        // the `.take` method is coming from the AsyncReadExt trait imported above
        let mut reader = input_file.take(chunk_size);
        let mut output = File::create(&output_path).await.unwrap();
        loop {
            let bytes_copied = tokio::io::copy(&mut reader, &mut output).await.unwrap();
            if bytes_copied == 0 {
                // EOF on the initial reader, remove the empty file we created
                break;
            }
            // our reader is now exhausted, but that doesn't mean the underlying reader
            // is. So we recover it, and we create the next chunked reader
            reader = reader.into_inner().take(chunk_size);
        }
        let file_size = tokio::fs::metadata(output_path).await.unwrap().len();
        assert!((chunk_size > 0 && file_size > 0) || chunk_size == 0 && file_size == 0);
    }
}
```

I think this is now resolved (for me).

Reply:
I think allowing a zero length is useful for cases where the length is computed dynamically. I've seen code along these lines:
```rs
let len = io.read_u64().await? as usize;
let mut buffer = Vec::with_capacity(len);
io.take(len).read_to_end(&mut buffer)?;
```
if there are length-zero messages, then the above makes use of `take(0)` behavior sometimes.